### ISSUE TITLE
Unsubscribe may block

### ISSUE THREAD

**Issue by mikix** [SoCo/SoCo#329]
I have encountered an issue cleaning up subscriptions when switching between networks.

I'm using soco on a phone. And when switching between wifi and 3G (or wifi and wifi), I clean old subscriptions before maybe making new ones.

Normally unsubscribe works fine. But in this case, it blocks for a long time. Presumably it is confused because all of a sudden it's on a new network with different ips and such.

I would not expect this behavior, so I'm assuming it's a bug. But it's also very possible I am driving soco wrong. Is there a best-practice for resetting state when switching networks?

**Comment by KennethNielsen** [comment]
Hallo @mikix

Thanks for reporting this. Which kind of subscriptions are we talking about, for events?

**Comment by mikix** [comment]
At the time of this bug, I was only dealing with subscriptions on renderingControl.

**Comment by KennethNielsen** [comment]
@mikix I think I can fix this by passing a timeout through to requests, I just need to read through the events code first to make sure I understand it.

One would assume, that the unit will delete the subscription, once it can no longer reach a client, so I would assume, that in case if timeout it is safe to simply proceed as if we had been unsubscribed.

Can you please check that the subscriptions are indeed deleted automatically on the unit, by subscribing (and getting event) on network 1, switch to network 2, then back to network 1 again and check that you are not still getting events?

**Comment by KennethNielsen** [comment]
Ok. I think I have a handle on how the code works now. I would still like you to perform the check above since I think there is an event listener running on the IP the device has when first importing the module. So in order to facilitate a network change, I think it will be necessary to reset it. Forget about un-subscribing for a moment, please try the following after your network change:

```python
from soco import events

### All you usual code that sets up subscriptions

### Then after network change do:
events.event_listener.stop()
time.sleep(10)
events._sid_to_event_queue.clear()
events._sid_to_service.clear()

# Then re-create your subscriptions
```

I'm not completely sure this is the best way to do it, I will look at it some more, but if we find a solution that works, we can create a convenience function for it.

**Comment by mikix** [comment]
As for your first test (subscribe on network 1, switch to 2, switch back to 1), I received events on network 1 the second time.  I don't think enough time passed for the unit to understand that I had left network 1.  So the result doesn't super surprise me.  Maybe staying on network 2 longer would have a different result.

Will try your second test.

**Comment by mikix** [comment]
Second test resulted in a long-block for the events.event_listener.stop() call.  (not the sleep call but the stop call itself blocked)

**Comment by KennethNielsen** [comment]
Ok. So this is real ugly, and will not be committed to main, just want to see if it works.

Can you try the reset_events branch on my SoCo archive: https://github.com/KennethNielsen/SoCo/tree/reset_events. In there, the events module has a ``reset_event_listener`` function. Try and call that and forget about the subscriptions. The function takes optional waittime arguments, which should be some number of seconds that it takes for everything to settle down after forcefully closing the socket and just deleting the old event listener. If you get errors, you may need to raise it, right now it is 4 seconds.

**Comment by mikix** [comment]
I'm happy to test that, but might not get to it today.

But I had a separate thought: from my app's perspective, cleaning up old subscriptions is ideally a fire-and-forget sort of thing.  Being unresponsive for even 4 seconds is too long.  Maybe a good approach is to spin off a thread to do the unsubscribe() calls and then just continue.  If the thread blocks for a couple minutes, that's fine.

**Comment by KennethNielsen** [comment]
@mikix hold of on testing.

With regards to un-subscribing in a separate thread, I don't think that will work. The event listener starts up a server for listening for notifications at import time and uses the IP address at import time as well. So in order for new subscriptions to work, it will have to be restarted. But your comment about shutting things down in another thread gave me an idea. I will hack at it a little more and send you something else.

**Comment by KennethNielsen** [comment]
Ok. I have a better implementation ready (which if it works, should be ok to go in main).

Please test the branch ``event_reset_once_more`` in my SoCo archive: https://github.com/KennethNielsen/SoCo.

With this I have added two changes.

 1. I have added a timeout option to ``unsubscribe`` on the subscriptions. When unsubscribing after a network change you just set a short timeout. The actual unsubscribing with the units will of course fail, but the internal bookkeeping of the subscriptions will be performed as if it had succeeded and we expect the units to clear that up themselves afterwards.
 2. I have added a ``https://github.com/KennethNielsen/SoCo`` module level function to events. It takes two arguments, the first is a waittime, in case you are resettings events on the same network and need to wait for the socket to be freed, the second is a boolean for whether to shut down the old event listener in a new thread. So for your usecase, you should just call it as: ``reset_event_listener(wait_after_shutdown=0, shutdown_in_thread=True)`` and it *should* do the job.

So to test, simply unsubscribe the old subscriptions with a (arbitrarily short) timeout and call the function to reset the listener. After this you should be ready to make new subscriptions again (and all if this with no blocking).

Let me know how it goes.

**Comment by mikix** [comment]
A) Consider moving the "global event_listener" to the beginning of "reset_event_listener" -- you get this warning otherwise:
events.py:644: SyntaxWarning: name 'event_listener' is used prior to global declaration

B) Consider noting in the docstring that timeout is in seconds.  Obvious if you know the implementation details / API of requests.  But still.

C) Consider having a is_running check inside of reset_event_listener.  I can do that in my code, but every single caller will need to make that check, it would be nice if the function was just a no-op in that case.  (If the caller doesn't do that check, they get a an AttributeError on NoneType for stop_flag.)

D) If I use shutdown_in_thread, I get an "OSError: [Errno 98] Address already in use" error.  Seems like a race between old event_listener and new one.

E) I think even in this design, I'd want to unsubscribe in a thread?  I can block for only tiny amounts (like 0.1s), but then I run the risk of cutting a request short.  And any blocking is bad.  Seems like I want the unsubscribe to be asynchronous regardless of how much time it takes.  But that's orthogonal to reset_event_listener.

**Comment by mikix** [comment]
Looking into what is actually blocking...  Seems that urlopen is one of the calls that is causing a block.  Adding timeout=0 to it makes it not block anymore, as you'd expect (I don't know if that's a practical value in other cases -- it's not like we care about an answer to that call though).

Then I hit the _listener_thread.join() call which blocks...  Surely there's a way for socketserver.TCPServer to notice when network is dead...

**Comment by mikix** [comment]
I can look into fixing this.  It's easier for me to test my scenario than you I bet.

**Comment by KennethNielsen** [comment]
@mikix thanks for the feedback. I have fixed the issues (and made a much simpler version really).

With regards to your comments:

E) This was the real problem. It is not a race per-se. But when you are done with a port, you need to jump through a few hoops to make sure it is free for future use and since we want to start a server on **the same** port, this had to be fixed.
E) the way I understand the code, un-subcribing in a thread will not solve anything. It is not the un-subscriptions that we were waiting on, it was the restarting of the server (see D). As already discussed, if you know that there is a network change, you know that the un-subscribe will block for a long time without achieving anything and so you can just set the timeout arbitrarily short (say 0.001) and it should actually unsubscribe faster than if we had still been connected. In the new version I have, I actually demand that all subscriptions has been cleared before stopping, so in this case, you cannot stop them in a thread. Just set the timeout really short.

With regards to:
>Looking into what is actually blocking... Seems that urlopen is one of the calls that is causing a block. Adding timeout=0 to it makes it not block anymore, as you'd expect (I don't know if that's a practical value in other cases -- it's not like we care about an answer to that call though).
>
>Then I hit the _listener_thread.join() call which blocks... Surely there's a way for socketserver.TCPServer to notice when network is dead...

Yes, that is what is blocking. The problem however is not that it is blocking, the problem is that the request never lands. Notice that the server is in loop where it is waiting for requests, and only after receiving each request will it check if it has been asked to shut down (``EventServerThread.run``). one way to fix this, is to set the ``handle_request`` up with a timeout. We do not want to do that for the general case though, as it makes the while run through an iteration every timeout, which is needless overhead, so therefore you will have to do it yourself.

I have made a new version in the same branch (currently at commit 90450ee98b). It is much simpler (less invasive to the code) and this now works for me:
```python

from __future__ import print_function
try:
    from queue import Empty
except:  # Py2.7
    from Queue import Empty

import soco
from pprint import pprint
from soco import events

# pick a device at random
device = soco.discover().pop()
print(device.player_name)
sub = device.renderingControl.subscribe()
events.set_event_listener_timeout(0.1)
print('Event server now running. Press ctrl-c to reset')

while True:
    try:
        event = sub.events.get(timeout=0.5)
        pprint(event.variables)
    except Empty:
        pass
    except KeyboardInterrupt:
        # If you suspect a network change, you can make this as short as you
        # want
        sub.unsubscribe(0.1)
        break

events.stop_event_listener()

sub = device.renderingControl.subscribe()
print('Event server now running again.')

while True:
    try:
        event = sub.events.get(timeout=0.5)
        pprint(event.variables)
    except Empty:
        pass
    except KeyboardInterrupt:
        sub.unsubscribe(0.1)
        break
```

The important parts are ``events.set_event_listener_timeout(0.1)``. Two important things:
 1. This call must be made AFTER making at least one subscription, so that the event server is running
 2. Must be called after each (re)-start of the event server (This could be fixed, if it turns out to be a large in-convenience, but first I want to see if it works.

Afterwards simple do `events.stop_event_listener()`, which will stop the event server. It will be automatically restarted when making a new subscription. NOTE that all the old subscriptions must be un-subscribed first. Use the timeout to do that fast.

Let me know how it goes.

**Comment by KennethNielsen** [comment]
Thanks btw for the offer to fix it yourself, I had, however, already spent quite a bit of time on it when I saw your message.

The solution above, provided that it works, should be something that can be turned into something more elegant. I believe most of the "do this before" and "do this after" things can be fixed

**Comment by mikix** [comment]
Hey, sorry I've been so silent. I've just been very busy. And am going to be busy through the weekend. I'll try to test this next week.

**Comment by KennethNielsen** [comment]
No rush from my side ;)

**Comment by mikix** [comment]
Some comments:
- Mostly works!
- Typo in "...subscription has been mad" error message (missing final e)
- Going wifi (with zones) to 3g (no zones found) works (yay!), but then going back to the original wifi and calling soco.discover() again returns no zones. I'm not sure why yet.
- I'm still unhappy with a solution that requires me to set a timeout for the event listener thread.  As you note, it's needless overhead.  I'm wondering what ways there are to interrupt the select() call that socketserver makes...

**Comment by mikix** [comment]
OK, ignore my issue with discover() returning no zones.  That was user error.

Thus, this branch works as advertised and solves my most pressing need of avoiding blocking code.  But I'm still interested in a solution that doesn't use timeouts in a loop.

**Comment by mikix** [comment]
(To be clear, I'm happy with leaving the avoid-timeout/"interrupt select() call in socketserver" improvement to a future branch.  This helps enough to land in my opinion.)

**Comment by lawrenceakka** [comment]
Sorry I'm a bit late to the party. The fact that I wrote the events code doesn't mean that I know how it works!

Sonos units do indeed continue trying to send events for some time after an unsuccessful attempt.  You can see this at http://{device_ip}:1400/status/opt/log/anacapa.trace.  Eventually they will give up and drop the subscription.

I don't see any real harm in setting a short timeout in the unsubscribe method, since we don't really care out whether the request is successful or not. It is possible that an error will be raised by the sonos device, but it will drop the subscription in due course anyway. I don't know that we need the timeout to be configurable, but again it doesn't do any harm.  (We need to be careful about all the timeout variables though - there are now quite a few and it might be confusing.)

At some point, we might go full async (perhaps using gevent, which finally supports Python 3, and  https://github.com/kennethreitz/grequests )

**Comment by lawrenceakka** [comment]
> I'm still unhappy with a solution that requires me to set a timeout for the event listener thread. As you note, it's needless overhead. I'm wondering what ways there are to interrupt the select() call that socketserver makes...

This is how SocketServer does it - see the comments at https://hg.python.org/cpython/file/2.7/Lib/SocketServer.py#l231

I wonder if forcibly closing the socket might work, e.g.:

```diff
diff --git a/soco/events.py b/soco/events.py
--- a/soco/events.py
+++ b/soco/events.py
@@ -263,11 +263,11 @@
         # Start the server on the local IP at port 1400 (default).
         # Handling of requests is delegated to instances of the
         # EventNotifyHandler class
-        listener = EventServer(self.address, EventNotifyHandler)
-        log.info("Event listener running on %s", listener.server_address)
+        self.listener = EventServer(self.address, EventNotifyHandler)
+        log.info("Event listener running on %s", self.listener.server_address)
         # Listen for events untill told to stop
         while not self.stop_flag.is_set():
-            listener.handle_request()
+            self.listener.handle_request()
 
 
 class EventListener(object):
@@ -318,7 +318,7 @@
         """Stop the event listener"""
         # Signal the thread to stop before handling the next request
         self._listener_thread.stop_flag.set()
-        # Send a dummy request in case the http server is currently listening
+        # # Send a dummy request in case the http server is currently listening
         try:
             urlopen(
                 'http://%s:%s/' % (self.address[0], self.address[1]))
@@ -327,6 +327,7 @@
             # which we ignore.
             pass
         # wait for the thread to finish
+        self._listener_thread.listener.socket.close()
         self._listener_thread.join()
         self.is_running = False
         log.info("Event listener stopped")
```

I am worried that this is not thread safe however (possible concurrent access of socket object)

**Comment by KennethNielsen** [comment]
Hi @lawrenceakka

> The fact that I wrote the events code doesn't mean that I know how it works!

;)

Just to be clear, I'm not proposing this for inclusion, this was more of an experiment and see if I could learn a bit more about how this works.

One thing that I was wondering about while working with it. You set it up, so that the server it shut down by sending it an dummy event. That is one of the things that are giving us problems here, because if the network connection is gone, this dummy event cannot be set. So I was wondering about a way to forcibly close the server, but before looking for it I was really curious if there was a specific reason you set it up like that.

**Comment by the01** [comment]
@KennethNielsen, I think I can venture a guess here. I have written some event queues in the past and I found it cleaner/smoother (however you want to call it) to add a closing event at the current end of the queue. Then the code would work off the remaining events until it encountered the dummy event. It is kind of like a soft shutdown - versus a hard shutdown, when interrupting the server in whatever it is doing.
Am I correct, @lawrenceakka ?

**Comment by KennethNielsen** [comment]
@the01 yes, that is of course how you would normally do it. I was just so locked in with this problem of forcing a shutdown that I didn't consider that.

**Comment by lawrenceakka** [comment]
What is happening in this code

```python
   while not self.stop_flag.is_set():
            listener.handle_request()
```
is that `handle_request` calls `socket.select()` behind the scenes, and that waits until the timeout before it returns. Sending a dummy http request to the socket address causes `select` to return earlier, and we don't need to wait for the timeout to expire. The problem of course is that in the scenario @mikix has,  the controller can no longer access the network that the socket is listening on, so the http call never gets through.

The alternatives are (1) to use a polling loop or, `serve_forever` (which is the same thing), or (2) to set a shorter timeout on the socket, or (3) to try and force the socket to close from outside its thread.

Lawrence

**Comment by KennethNielsen** [comment]
@lawrenceakka I'm unsure. What I was wondering, and what I think would be the most elegant solution if it is possibly, is if we are able to fake the dummy request. I.e. does the Server object has some secret methods that we could call to make it think it received the closing event, without actually having it go over the network. I'm unsure if this is possible and I had little luck googling for it.

**Comment by the01** [comment]
Since ```select()``` accepts an array of sockets (or objects with certain properties/functions), you could add a thread-safe dummy socket/object/whatever to the select call and write your closing event to that.
That would have the additional benefit of being able to feed your own data as if it was coming in over the wire :wink:

**Comment by mikix** [comment]
@the01 agreed, though socketserver doesn't give an easy way to edit the list.  I suppose the best way would be to just re-implement the handle_request call and write the select line ourselves.  By stuffing our own socket in there, we wouldn't need to rely on the network at all or harsh ways of aborting the select.
